### PR TITLE
“Did you mean?” hints for `choice()` value typos

### PR DESCRIPTION
**Pull request by dahlia**
Summary
-------

 -  `choice()` gains a `suggest` option that appends a “Did you mean?” hint when the input is close to a valid choice.
 -  A new `appendValueHint()` helper, exported from `@optique/core/suggestion`, lets custom value parsers add the same hint without reimplementing distance logic.
 -  The default is `"never"`, so existing error messages are unchanged.

Motivation
----------

When a user mistypes an option name, Optique already offers a “Did you mean?” hint. The same machinery was already public, but it was never wired into value-level errors. A `--mode devo` typo got the same dry “not one of: dev, prod” message it would have gotten without the library, even though `dev` is one edit away.

The `suggest` option closes that gap. It's opt-in for backward compatibility and supports four forms: the built-in nearest-match heuristic with default or custom thresholds, and a function for domain-specific logic.

Test plan
---------

 -  [ ] `mise test` passes (Deno, Node.js, Bun)
 -  [ ] `suggest: "nearest"` appends a hint for near-miss inputs
 -  [ ] Default (`suggest` omitted or `"never"`) produces the same error as before
 -  [ ] Invalid `suggest` values (e.g. `true`, `"nearset"`) throw `TypeError` at construction
 -  [ ] `appendValueHint()` returns the base message unchanged when no candidate is close enough
 -  [ ] Docs build passes (`pnpm build` in *docs/*)

### CODE REVIEW

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/valueparser.ts [hunk 1]
```diff
@@ -644,6 +676,22 @@ export function choice<const T extends string | number>(
     }
   }
   const stringInvalidChoice = stringOptions.errors?.invalidChoice;
+  const stringSuggest = stringOptions.suggest;
+  if (stringSuggest !== undefined) {
+    const isValidString = stringSuggest === "nearest" ||
+      stringSuggest === "never";
+    const isValidObject = typeof stringSuggest === "object" &&
+      stringSuggest !== null;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Addressed in 377ece65: added Array.isArray check to reject arrays, and the error message now reports "array" rather than the misleading "object" when an array is passed.

File: packages/core/src/valueparser.ts [hunk 2]
```diff
@@ -803,13 +852,31 @@ function formatStringChoiceError(
...
+  if (suggest == null || suggest === "never") return base;
+
+  if (suggest === "nearest") {
+    return appendValueHint(base, input, choices);
   }
-  return formatDefaultChoiceError(input, choices);
+
+  if (typeof suggest === "function") {
+    const hints = suggest(input, choices);
+    if (!hints || hints.length === 0) return base;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Addressed in 377ece65: changed the guard to Array.isArray(hints), so a string return from the custom function is treated as no hints rather than being iterated character by character.

File: docs/cookbook.md [hunk 3]
```diff
@@ -1373,6 +1373,108 @@ The `passThrough()` parser supports three capture formats:
 > The `"greedy"` format can shadow explicit parsers. Place it carefully,
 > typically as the last field in a subcommand-specific `object()`.
 
+### “Did you mean?” for value typos
+
+*This feature is available since Optique 1.2.0.*
+
+When a user mistypes an option value — `--mode devo` instead of `dev` — the
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/cookbook.md]: Addressed in b3c70b3c.

File: packages/core/src/suggestion.test.ts [hunk 4]
```diff
@@ -1041,3 +1042,52 @@ describe("property-based tests", () => {
...
+    const result = appendValueHint(base, "xyz", ["dev", "prod"]);
+    assert.deepEqual(result, base);
+  });
+
+  it("should append Did you mean hint when input is close to a candidate", () => {
+    const base = message`Expected one of dev or prod, but got devo.`;
+    const result = appendValueHint(base, "devo", ["dev", "prod"]);
+    const str = formatMessage(result);
+    assert.ok(str.includes("Did you mean"), `Expected hint in: ${str}`);
+    assert.ok(str.includes("dev"), `Expected "dev" in: ${str}`);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.test.ts]: Addressed in 1aa50a37: changed the base message to "Invalid environment." (no candidate names) and the assertion now checks for the backtick-wrapped form `dev` so it verifies the hint was actually appended.

File: packages/core/src/suggestion.ts [hunk 5]
```diff
@@ -425,6 +425,54 @@ export function createErrorWithSuggestions(
...
+  const suggestions = findSimilar(
+    input,
+    candidates,
+    options != null
+      ? { ...DEFAULT_FIND_SIMILAR_OPTIONS, ...options }
+      : undefined,
+  );
+  const suggestionMsg = createSuggestionMessage(suggestions);
+  return suggestionMsg.length > 0
+    ? [...base, lineBreak(), lineBreak(), ...suggestionMsg]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts]: Addressed in 1aa50a37: the separators are now gated on base.length > 0, so an empty base returns just the hint without leading newlines.

File: packages/core/src/valueparser.test.ts [hunk 6]
```diff
@@ -1727,6 +1727,160 @@ describe("choice", () => {
...
+        TypeError,
+      );
+    });
+
+    it("should throw TypeError for suggest: 0", () => {
+      assert.throws(
+        () => choice(["dev", "prod"], { suggest: 0 as never }),
+        TypeError,
+      );
+    });
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:1881]: Addressed in 377ece65: all three TypeError tests now assert both the error name and the message content via /Expected suggest to be/i.

File: packages/core/src/valueparser.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Addressed in 377ece65: when the object form is used, maxDistance and maxSuggestions are now validated to be finite numbers at construction time, throwing TypeError before the value ever reaches findSimilar().

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/valueparser.ts [hunk 7]
```diff
@@ -803,13 +880,31 @@ function formatStringChoiceError(
...
+    return appendValueHint(base, input, choices);
   }
-  return formatDefaultChoiceError(input, choices);
+
+  if (typeof suggest === "function") {
+    const hints = suggest(input, choices);
+    if (!Array.isArray(hints) || hints.length === 0) return base;
+    const suggestionMsg = createSuggestionMessage(hints);
+    return suggestionMsg.length > 0
+      ? [...base, lineBreak(), lineBreak(), ...suggestionMsg]
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Declined: lineBreak is already imported at line 3 of valueparser.ts (`import { lineBreak, ... } from "./message.ts"`). No change needed.

File: packages/core/src/suggestion.ts [hunk 8]
```diff
@@ -425,6 +425,55 @@ export function createErrorWithSuggestions(
...
+    Pick<FindSimilarOptions, "maxDistance" | "maxSuggestions">
+  >,
+): Message {
+  const suggestions = findSimilar(
+    input,
+    candidates,
+    options != null
+      ? { ...DEFAULT_FIND_SIMILAR_OPTIONS, ...options }
+      : undefined,
+  );
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:509]: Addressed in 46598922: switched to explicit ?? fallbacks for each field. Also took the opportunity to switch from optionName terms to value terms in appendValueHint(), so candidates are JSON-escaped correctly (see Codex comment below).

File: packages/core/src/valueparser.ts [hunk 9]
```diff
@@ -644,6 +676,50 @@ export function choice<const T extends string | number>(
...
+      if (
+        obj.maxSuggestions !== undefined &&
+        (typeof obj.maxSuggestions !== "number" ||
+          !isFinite(obj.maxSuggestions))
+      ) {
+        throw new TypeError(
+          `Expected suggest.maxSuggestions to be a finite number, but got ` +
+            `${typeof obj.maxSuggestions}: ${String(obj.maxSuggestions)}.`,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:723]: Addressed in ff2594d8: validation now requires non-negative integers for maxDistance and positive integers for maxSuggestions, with updated error messages and boundary-case tests.

File: packages/core/src/suggestion.ts [hunk 10]
```diff
@@ -425,6 +425,55 @@ export function createErrorWithSuggestions(
...
+  const suggestionMsg = createSuggestionMessage(suggestions);
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts]: Addressed in 46598922: appendValueHint() now builds the hint using value terms instead of going through createSuggestionMessage(). value terms use JSON.stringify, so a candidate like foo`bar renders as "foo`bar" rather than breaking the backtick pairing.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/suggestion.ts [hunk 11]
```diff
@@ -425,6 +425,70 @@ export function createErrorWithSuggestions(
...
+  );
+  if (suggestions.length === 0) return base;
+  let suggestionMsg: Message;
+  if (suggestions.length === 1) {
+    suggestionMsg = message`Did you mean ${value(suggestions[0])}?`;
+  } else {
+    const parts: MessageTerm[] = [text("Did you mean one of these?")];
+    for (const suggestion of suggestions) {
+      parts.push(text("\n  "), value(suggestion));
+    }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts]: Addressed in e16654a6: the multi-suggestion loop now uses lineBreak(), text("  ") instead of text("\n  "), so each hint renders on its own line.

File: packages/core/src/valueparser.ts [hunk 12]
```diff
@@ -803,13 +883,38 @@ function formatStringChoiceError(
...
+    const hints = suggest(input, choices);
+    if (!Array.isArray(hints) || hints.length === 0) return base;
+    let suggestionMsg: Message;
+    if (hints.length === 1) {
+      suggestionMsg = message`Did you mean ${valueTerm(hints[0])}?`;
+    } else {
+      const parts: MessageTerm[] = [text("Did you mean one of these?")];
+      for (const hint of hints) {
+        parts.push(text("\n  "), valueTerm(hint));
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Addressed in e16654a6: the function-form branch now calls appendValueSuggestions(), which uses lineBreak(), text("  ") for multi-hint lists.

File: packages/core/src/suggestion.ts [hunk 13]
```diff
@@ -425,6 +425,70 @@ export function createErrorWithSuggestions(
...
+    suggestionMsg = parts;
+  }
+  return base.length > 0
+    ? [...base, lineBreak(), lineBreak(), ...suggestionMsg]
+    : suggestionMsg;
+}
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/suggestion.ts:511]: Addressed in e16654a6: extracted appendValueSuggestions() as a shared helper in suggestion.ts; appendValueHint() delegates to it, and so does the function-form suggest branch in valueparser.ts, fixing both the duplicate logic and the missing empty-base guard.

File: packages/core/src/valueparser.ts [hunk 14]
```diff
@@ -6,12 +6,17 @@ import {
...
 } from "./message.ts";
 import { isDerivedValueParser } from "./internal/dependency.ts";
 import { ensureNonEmptyString, type NonEmptyString } from "./nonempty.ts";
 import type { Mode, ModeIterable, ModeValue, Suggestion } from "./parser.ts";
-import { deduplicateSuggestions } from "./suggestion.ts";
+import {
+  appendValueHint,
+  deduplicateSuggestions,
+  type FindSimilarOptions,
+} from "./suggestion.ts";
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:19]: Addressed in e16654a6: appendValueSuggestions is now imported from ./suggestion.ts.

File: packages/core/src/valueparser.ts [hunk 15]
```diff
@@ -803,13 +883,38 @@ function formatStringChoiceError(
...
+      suggestionMsg = message`Did you mean ${valueTerm(hints[0])}?`;
+    } else {
+      const parts: MessageTerm[] = [text("Did you mean one of these?")];
+      for (const hint of hints) {
+        parts.push(text("\n  "), valueTerm(hint));
+      }
+      suggestionMsg = parts;
+    }
+    return [...base, lineBreak(), lineBreak(), ...suggestionMsg];
   }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:911]: Addressed in e16654a6: the function-form branch now simply calls appendValueSuggestions(base, hints), which handles the empty-base guard and the correct rendering.

File: packages/core/src/valueparser.ts [hunk 16]
```diff
@@ -803,13 +883,38 @@ function formatStringChoiceError(
...
+    if (hints.length === 1) {
+      suggestionMsg = message`Did you mean ${valueTerm(hints[0])}?`;
+    } else {
+      const parts: MessageTerm[] = [text("Did you mean one of these?")];
+      for (const hint of hints) {
+        parts.push(text("\n  "), valueTerm(hint));
+      }
+      suggestionMsg = parts;
+    }
+    return [...base, lineBreak(), lineBreak(), ...suggestionMsg];
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts]: Addressed in e16654a6: the function-form branch now calls appendValueSuggestions(base, hints), which guards on base.length before prepending the lineBreaks.

**Comment by dahlia** [comment]
@codex review

File: docs/concepts/valueparsers.md [hunk 17]
```diff
@@ -654,6 +654,53 @@ const depth = choice([8, 10, 12]);
...
+const mode = choice(["dev", "prod"], { suggest: "nearest" });
+~~~~
+
+With this option, the same error becomes:
+
+~~~~ bash
+$ myapp --mode devo
+Error: Expected one of dev and prod, but got devo.
+
+Did you mean `dev`?
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/concepts/valueparsers.md]: Addressed in 694ee20a: the example output now shows "dev" (double quotes) to match what value() terms render.

File: docs/cookbook.md [hunk 18]
```diff
@@ -1373,6 +1373,108 @@ The `passThrough()` parser supports three capture formats:
...
+const mode = choice(["dev", "staging", "prod"], { suggest: "nearest" });
+~~~~
+
+A typo now produces:
+
+~~~~ bash
+$ myapp --mode devo
+Error: Expected one of dev, staging, and prod, but got devo.
+
+Did you mean `dev`?
```
[the remark this answers is not in the document]
Comment (dahlia) [reply docs/cookbook.md]: Addressed in 694ee20a: updated to show "dev" instead of `dev`.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/valueparser.ts [hunk 19]
```diff
@@ -644,6 +676,53 @@ export function choice<const T extends string | number>(
     }
   }
   const stringInvalidChoice = stringOptions.errors?.invalidChoice;
+  const stringSuggest = stringOptions.suggest;
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.ts:679]: Addressed in 23a526b7: after validation, the object fields are copied into a fresh object so the closure captures stable values rather than the caller's live reference.

**Comment by dahlia** [comment]
@codex review

File: packages/core/src/valueparser.test.ts [hunk 20]
```diff
@@ -1727,6 +1727,215 @@ describe("choice", () => {
...
+      (suggestObj as Record<string, unknown>).maxDistance = 0;
+      const result = parser.parse("devo");
+      assert.ok(!result.success);
+      if (!result.success) {
+        const str = formatMessage(result.error);
+        assert.ok(
+          str.includes("Did you mean"),
+          `Hint should appear with original maxDistance: ${str}`,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (dahlia) [reply packages/core/src/valueparser.test.ts:1935]: Declined: the test is checking behavioral correctness (does the snapshot preserve the hint after mutation?) rather than message structure. The formatMessage + includes pattern is consistent with adjacent tests in this describe block (the custom-hint and suppress-hint cases). The deepStrictEqual example in the suggestion also references optionName terms, which are not part of a Did you mean? hint value message.
